logging/pyansys_logging.py

```diff
@@ -0,0 +1,498 @@
+        if defaults:
+            values = defaults | record.__dict__
+        else:
+            values = record.__dict__
+
+        # Here we can make any changes we want in the record, for
+        # example adding a key.
+
+        # We could create an if here if we want conditional formatting, and even
+        # change the record.__dict__.
+        # Since now we don't want to create conditional fields, it is fine to keep
+        # the same MSG_FORMAT for all of them.
+
+        # For the case of logging exceptions to the logger.
+        values.setdefault("instance_name", "")
+
+        return STDOUT_MSG_FORMAT % values
+
+
+class PyProjectFormatter(logging.Formatter):
+    """Customized ``Formatter`` class used to overwrite the defaults format styles."""
+
+    def __init__(
+        self,
+        fmt=STDOUT_MSG_FORMAT,
+        datefmt=None,
+        style="%",
+        validate=True,
+        defaults=None,
+    ):
+        if sys.version_info[1] < 8:
+            super().__init__(fmt, datefmt, style)
+        else:
+            # 3.8: The validate parameter was added
+            super().__init__(fmt, datefmt, style, validate)
+        self._style = PyAnsysPercentStyle(fmt, defaults=defaults)  # overwriting
+
+
+class InstanceFilter(logging.Filter):
+    """Ensures that instance_name record always exists."""
+
+    def filter(self, record):
+        if not hasattr(record, "instance_name"):
+            record.instance_name = ""
+        return True
+
+
+class Logger:
+    """Logger used for each PyProject session.
+
+    This class allows you to add a handler to a file or standard output.
+
+    Parameters
+    ----------
+    level : int, optional
+        Logging level to filter the message severity allowed in the logger.
+        The default is ``logging.DEBUG``.
+    to_file : bool, optional
+        Write log messages to a file. The default is ``False``.
+    to_stdout : bool, optional
+        Write log messages into the standard output. The
+        default is ``True``.
+    filename : str, optional
+        Name of the file where log messages are written to.
+        The default is ``None``.
+    """
+
+    file_handler = None
+    std_out_handler = None
+    _level = logging.DEBUG
+    _instances = {}
+
+    def __init__(
+        self, level=logging.DEBUG, to_file=False, to_stdout=True, filename=FILE_NAME
+    ):
+        """Customized logger class for PyProject.
+
+        Parameters
+        ----------
+        level : str, optional
+            Level of logging as defined in the package ``logging``. By
+            default ``'DEBUG'``.
+        to_file : bool, optional
+            To record the logs in a file, by default ``False``.
+        to_stdout : bool, optional
+            To output the logs to the standard output, which is the
+            command line. By default ``True``.
+        filename : str, optional
+            Name of the output file. By default ``"PyProject.log"``.
+        """
```

Comment:
BTW, don't document `__init__` any more than:
```suggestion
        """Initialize Logger class."""
```